cloudsmith_cli/core/api/init.py

```diff
@@ -40,15 +44,51 @@ def initialise_api(
     config.verify_ssl = ssl_verify
     config.client_side_validation = False
 
+    access_token = keyring.get_access_token(config.host)
+    if access_token:
+        auth_header = config.headers.get("Authorization")
+
+        # overwrite auth header if empty or is basic auth without username or password
+        if not auth_header or auth_header == config.get_basic_auth_token():
+            refresh_token = keyring.get_refresh_token(config.host)
+
+            try:
+                if keyring.should_refresh_access_token(config.host):
+                    new_access_token, new_refresh_token = saml.refresh_access_token(
+                        config.host, access_token, refresh_token
+                    )
+                    keyring.store_sso_tokens(
+                        config.host, new_access_token, new_refresh_token
+                    )
+            except ApiException:
+                keyring.update_refresh_attempted_at(config.host)
+
+                # try falling back to API key auth if refresh fails
+                if key:
+                    config.api_key["X-Api-Key"] = key
```

Comment:
❓ Should we communicate back to the user here that their session might've expired? The AWS CLI returns something like
```CredentialsProviderError: The SSO session associated with this profile has expired. To refresh this SSO session run aws sso login with the corresponding profile```

Ignore if a similar message is returned already from another part of the code.